quinn-proto/src/connection/cid_state.rs

```diff
@@ -0,0 +1,205 @@
+    pub(crate) fn track_lifetime(&mut self, new_cid_seq: u64, now: Instant) {
+        let lifetime = match self.cid_lifetime {
+            Some(lifetime) => lifetime,
+            None => return,
+        };
+        let expire_timestamp = now.checked_add(lifetime);
+        if let Some(expire_at) = expire_timestamp {
+            let last_record = self.retire_timestamp.back_mut();
+            if let Some(last) = last_record {
+                // Compare the timestamp with the last inserted record
+                // Combine into a single batch if timestamp of current cid is same as the last record
+                if expire_at == last.timestamp {
+                    debug_assert!(new_cid_seq > last.sequence);
+                    last.sequence = new_cid_seq;
+                    return;
+                }
+            }
+            self.retire_timestamp.push_back(CidTimeStamp {
+                sequence: new_cid_seq,
+                timestamp: expire_at,
+            });
+        }
+    }
+
+    /// Update local CID state when previously issued CID is retired
+    /// Return a flag that indicates whether a new CID needs to be pushed that notifies remote peer to respond `RETIRE_CONNECTION_ID`
+    pub(crate) fn on_cid_timeout(&mut self) -> bool {
+        // Whether the peer hasn't retired all the CIDs we asked it to yet
+        let unretired_ids_found =
+            (self.prev_retire_seq..self.retire_seq).any(|seq| self.active_seq.contains(&seq));
+        // According to RFC:
+        // Endpoints SHOULD NOT issue updates of the Retire Prior To field
+        // before receiving RETIRE_CONNECTION_ID frames that retire all
+        // connection IDs indicated by the previous Retire Prior To value.
+        // https://tools.ietf.org/html/draft-ietf-quic-transport-29#section-5.1.2
+        //
+        // All Cids are retired, `prev_retire_cid_seq` can be assigned to `retire_cid_seq`
+        if !unretired_ids_found {
+            self.prev_retire_seq = self.retire_seq;
+        }
+
+        let next_retire_sequence = self
+            .retire_timestamp
+            .pop_front()
+            .map(|seq| seq.sequence + 1);
+        let current_retire_prior_to = self.retire_seq;
+
+        // Advance `retire_cid_seq` if next cid that needs to be retired exists
+        if let Some(next_retire_prior_to) = next_retire_sequence {
+            if !unretired_ids_found && next_retire_prior_to > current_retire_prior_to {
+                self.retire_seq = next_retire_prior_to;
+            }
+        }
+
+        // Check if retirement of all CIDs that reach their lifetime is still needed
+        // If yes (return true), a new CID must be pushed with updated `retire_prior_to` field to remote peer.
+        // If no (return false), it means remote peer has proactively retired those CIDs (for other reasons) before CID lifetime is reached.
+        (current_retire_prior_to..self.retire_seq).any(|seq| self.active_seq.contains(&seq))
```

Comment:
This doesn't appear to depend in any way on the actual timestamps. What prevents this from issuing retire_prior_to for every single active CID, including those which should not yet be retired?

Reply:
I am not sure if i catch your point.

The return of this call determines whether the peer wants to publish a new CID with updated `retire_prior_to` field. This is allowed in RFC described as follows

```
 An endpoint MUST NOT
   provide more connection IDs than the peer's limit.  An endpoint MAY
   send connection IDs that temporarily exceed a peer's limit if the
   NEW_CONNECTION_ID frame also requires the retirement of any excess,
   by including a sufficiently large value in the Retire Prior To field.
```
When there is no active CID to retire (after we update the `retire_prior_to`  in this call), pushing a new CID may violate the above RFC statement. That's why a condition check is added here.

Reply:
I think I was confused about when this method was called. On review I don't see anything wrong here.